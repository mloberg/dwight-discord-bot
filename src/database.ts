import Keyv from 'keyv';
import { KeyvFile } from 'keyv-file';

import config from './config';
import logger from './logger';

export class Database {
    private impl?: Keyv;

    constructor(private readonly uri: string | undefined, private log: typeof logger) {}

    async get<T>(key: string, fallback?: T): Promise<T> {
        return (await this.kv.get(key)) || fallback;
    }

    async set<T>(key: string, value: T, ttl?: number): Promise<void> {
        await this.kv.set(key, value, ttl);
    }

    async delete(key: string): Promise<void> {
        await this.kv.delete(key);
    }

    async clear(): Promise<void> {
        await this.kv.clear();
    }

    private get kv() {
        if (this.impl) {
            return this.impl;
        }

        this.impl = this.uri?.startsWith('file:')
            ? new Keyv({
                  store: new KeyvFile({
                      filename: this.uri.slice(5),
                  }),
              })
            : new Keyv(config.dbUrl);

        this.impl.on('error', this.log.error);

        return this.impl;
    }
}

export default new Database(config.dbUrl, logger);
