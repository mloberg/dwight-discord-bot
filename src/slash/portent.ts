import db from '../db';
import { FriendlyError } from '../error';
import { SlashCommand } from '../types';
import { roll } from '../utils';

const portent: SlashCommand = {
    name: 'portent',
    description: 'Manage portent dice for a Divination Wizard',
    options: [
        {
            name: 'show',
            description: 'Show available portent dice',
            type: 'SUB_COMMAND',
        },
        {
            name: 'roll',
            description: 'Roll new portent dice',
            type: 'SUB_COMMAND',
            options: [
                {
                    name: 'one',
                    description: 'Value of first die',
                    type: 'INTEGER',
                },
                {
                    name: 'two',
                    description: 'Value of second die',
                    type: 'INTEGER',
                },
                {
                    name: 'three',
                    description: 'Value of third die (greater portent)',
                    type: 'INTEGER',
                },
                {
                    name: 'greater',
                    description: 'Roll three dice instead of two',
                    type: 'BOOLEAN',
                },
            ],
        },
        {
            name: 'use',
            description: 'Use a portent die',
            type: 'SUB_COMMAND',
            options: [
                {
                    name: 'dice',
                    description: 'Die value to use',
                    type: 'INTEGER',
                    required: true,
                },
            ],
        },
    ],
    async run(command) {
        const sub = command.options.getSubcommand();
        const key = `${command.guild?.id}-${command.user.id}`;
        const dice: number[] = (await db.get(key)) || [];

        if (sub === 'show') {
            await command.reply(dice.length ? dice.join(', ') : 'No available portent dice.');
            return;
        }

        if (sub === 'roll') {
            const results = [
                command.options.getInteger('one') || roll('d20'),
                command.options.getInteger('two') || roll('d20'),
            ];
            if (command.options.getBoolean('greater')) {
                results.push(command.options.getInteger('three') || roll('d20'));
            }
            await db.set(key, results);
            await command.reply(results.join(', '));
            return;
        }

        const used = command.options.getInteger('dice', true);
        const index = dice.indexOf(used);
        if (index === -1) {
            throw new FriendlyError(`No portent dice for ${used}. Available: ${dice.join(', ')}`);
        }
        dice.splice(index, 1);
        await db.set(key, dice);

        await command.reply(`Remaining: ${dice.join(', ')}`);
    },
};

export default portent;
