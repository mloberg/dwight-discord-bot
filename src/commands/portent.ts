import { SlashCommandBuilder } from '@discordjs/builders';
import { CommandInteraction } from 'discord.js';

import database from '../database';
import { FriendlyError } from '../error';
import { roll, rollOption } from '../utils';

export default {
    config: new SlashCommandBuilder()
        .setName('portent')
        .setDescription('Manage portent dice for a Divination Wizard')
        .addSubcommand((option) => option.setName('show').setDescription('Show available portent dice'))
        .addSubcommand((option) =>
            option
                .setName('roll')
                .setDescription('Roll new portent dice')
                .addBooleanOption((option) =>
                    option.setName('greater').setDescription('Roll three dice instead of two'),
                )
                .addIntegerOption(rollOption('d20').setName('one'))
                .addIntegerOption(rollOption('d20').setName('two'))
                .addIntegerOption(rollOption('d20').setName('three')),
        )
        .addSubcommand((option) =>
            option
                .setName('use')
                .setDescription('Use a portent die')
                .addIntegerOption(rollOption('d20').setName('dice').setRequired(true)),
        ),
    async handle(command: CommandInteraction): Promise<void> {
        const sub = command.options.getSubcommand();
        const key = `${command.guild?.id}-${command.user.id}`;
        const dice: number[] = await database.get(key, []);

        if (sub === 'show') {
            await command.reply(dice.length > 0 ? dice.join(', ') : 'No available portent dice.');
            return;
        }

        if (sub === 'roll') {
            const results = [
                command.options.getInteger('one') || roll('d20'),
                command.options.getInteger('two') || roll('d20'),
            ];
            if (command.options.getBoolean('greater')) {
                results.push(command.options.getInteger('three') || roll('d20'));
            }
            await database.set(key, results);
            await command.reply(results.join(', '));
            return;
        }

        const used = command.options.getInteger('dice', true);
        const index = dice.indexOf(used);
        if (index === -1) {
            throw new FriendlyError(`No portent dice for ${used}. Available: ${dice.join(', ')}`);
        }
        dice.splice(index, 1);
        await database.set(key, dice);

        await command.reply(`Remaining: ${dice.join(', ')}`);
    },
};
