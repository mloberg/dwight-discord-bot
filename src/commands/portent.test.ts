import database from '../database';
import { FriendlyError } from '../error';
import portent from './portent';

jest.mock('../database');

const mockDatabase = jest.mocked(database);

describe('/portent', () => {
    beforeEach(() => {
        mockDatabase.get.mockReset();
        mockDatabase.set.mockReset();
    });

    it('is a slash command', () => {
        expect(portent).toMatchSnapshot();
    });

    describe('show', () => {
        it('shows available portent dice', async () => {
            mockDatabase.get.mockResolvedValue([1, 20]);

            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('show');

            await portent.handle(command);
            expect(mockDatabase.get).toHaveBeenCalledWith('1234-6789', []);
            expect(command.reply).toHaveBeenCalledWith('1, 20');
        });

        it('shows no available dice', async () => {
            mockDatabase.get.mockResolvedValue([]);

            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('show');

            await portent.handle(command);
            expect(command.reply).toHaveBeenCalledWith('No available portent dice.');
        });
    });

    describe('roll', () => {
        it('generates portent dice', async () => {
            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('roll');

            await portent.handle(command);
            expect(command.reply).toHaveBeenCalledWith(expect.stringMatching(/^\d+, \d+$/));
            expect(mockDatabase.set).toHaveBeenCalledWith('1234-6789', expect.any(Array));
        });

        it('uses given portent dice', async () => {
            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('roll');
            command.options.getInteger.mockReturnValueOnce(1);
            command.options.getInteger.mockReturnValueOnce(2);
            // does not use unless greater portent
            command.options.getInteger.mockReturnValueOnce(3);

            await portent.handle(command);
            expect(command.reply).toHaveBeenCalledWith('1, 2');
            expect(mockDatabase.set).toHaveBeenCalledWith('1234-6789', [1, 2]);
        });

        it('generates greater portent dice', async () => {
            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('roll');
            command.options.getBoolean.mockReturnValue(true);

            await portent.handle(command);
            expect(command.reply).toHaveBeenCalledWith(expect.stringMatching(/^\d+, \d+, \d+$/));
        });

        it('uses given greater portent dice', async () => {
            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('roll');
            command.options.getBoolean.mockReturnValue(true);
            command.options.getInteger.mockReturnValueOnce(1);
            command.options.getInteger.mockReturnValueOnce(2);
            command.options.getInteger.mockReturnValueOnce(3);

            await portent.handle(command);
            expect(command.reply).toHaveBeenCalledWith('1, 2, 3');
            expect(mockDatabase.set).toHaveBeenCalledWith('1234-6789', [1, 2, 3]);
        });
    });

    describe('use', () => {
        it('removes a portent die', async () => {
            mockDatabase.get.mockResolvedValue([1, 5, 20]);

            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('use');
            command.options.getInteger.mockReturnValue(5);

            await portent.handle(command);
            expect(mockDatabase.set).toHaveBeenCalledWith('1234-6789', [1, 20]);
            expect(command.reply).toHaveBeenCalledWith('Remaining: 1, 20');
        });

        it('removes one portent die if two match', async () => {
            mockDatabase.get.mockResolvedValue([20, 20]);

            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('use');
            command.options.getInteger.mockReturnValue(20);

            await portent.handle(command);
            expect(mockDatabase.set).toHaveBeenCalledWith('1234-6789', [20]);
            expect(command.reply).toHaveBeenCalledWith('Remaining: 20');
        });

        it('throws error if no portent die matches', async () => {
            mockDatabase.get.mockResolvedValue([1]);

            const command = createMockCommand({ guild: { id: '1234' }, user: { id: '6789' } });
            command.options.getSubcommand.mockReturnValue('use');
            command.options.getInteger.mockReturnValue(20);

            await expect(portent.handle(command)).rejects.toMatchError(
                new FriendlyError('No portent dice for 20. Available: 1'),
            );
        });
    });
});
