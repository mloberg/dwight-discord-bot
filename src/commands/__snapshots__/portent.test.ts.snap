// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/portent is a slash command 1`] = `
Object {
  "config": Object {
    "default_permission": undefined,
    "description": "Manage portent dice for a Divination Wizard",
    "name": "portent",
    "options": Array [
      Object {
        "description": "Show available portent dice",
        "name": "show",
        "options": Array [],
        "type": 1,
      },
      Object {
        "description": "Roll new portent dice",
        "name": "roll",
        "options": Array [
          Object {
            "description": "Roll three dice instead of two",
            "name": "greater",
            "required": false,
            "type": 5,
          },
          Object {
            "choices": undefined,
            "description": "Value of a d20 roll.",
            "name": "one",
            "required": false,
            "type": 4,
          },
          Object {
            "choices": undefined,
            "description": "Value of a d20 roll.",
            "name": "two",
            "required": false,
            "type": 4,
          },
          Object {
            "choices": undefined,
            "description": "Value of a d20 roll.",
            "name": "three",
            "required": false,
            "type": 4,
          },
        ],
        "type": 1,
      },
      Object {
        "description": "Use a portent die",
        "name": "use",
        "options": Array [
          Object {
            "choices": undefined,
            "description": "Value of a d20 roll.",
            "name": "dice",
            "required": true,
            "type": 4,
          },
        ],
        "type": 1,
      },
    ],
  },
  "handle": [Function],
}
`;
